### PR TITLE
[ci] Add mage command to compare bench reports

### PR DESCRIPTION
**Pull request by marc-gr**
## What does this PR do?

Adds a mage target that will compare two given benchmark results folders.

The source folder will be compared against a target folder, and for any benchmark result in source it will generate a report, showing the variation between the two.

Example report:

#### Package `google_workspace`

Data stream | Previous EPS | New EPS | Diff (%) | Result
----------- | ------------ | ------- | -------- | ------
`admin` | 0 | 2906.13 | 2906.13 (100%) | :+1:
`drive` | 0 | 4882.81 | 4882.81 (100%) | :+1:
`groups` | 0 | 5232.86 | 5232.86 (100%) | :+1:
`login` | 0 | 4972.65 | 4972.65 (100%) | :+1:
`saml` | 0 | 6377.55 | 6377.55 (100%) | :+1:
`user_accounts` | 0 | 12626.26 | 12626.26 (100%) | :+1:

#### Package `windows`

Data stream | Previous EPS | New EPS | Diff (%) | Result
----------- | ------------ | ------- | -------- | ------
`forwarded` | 5854.8 | 3148.61 | -2706.19 (-85.95%) | :green_heart:
`powershell` | 5854.8 | 5263.16 | -591.64 (-11.24%) | :green_heart:
`powershell_operational` | 5854.8 | 7137.76 | 1282.96 (17.97%) | :broken_heart:
`sysmon_operational` | 5854.8 | 7107.32 | 1252.52 (17.62%) | :broken_heart:

----

The command is used as such:

`mage BenchReport "./build/benchmark-results" "/main/build/benchmark-results" 10 report.md`

The first argument is the source folder (the newest results), the second argument is the target folder (the previous results we want to compare against), the third argument is the thresholds, it indicates an expected % variance threshold where we assume nothing has changed and might be caused due to run-to-run factors, and the latest argument is the file name where the report will be saved.

The intention of this command is to be used in CI to generate useful information in PRs.

### CODE REVIEW

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -150,6 +155,48 @@ pipeline {
   }
 }
 
+def gitHubCommentWithBenchmarkReport() {
+  // report only on a PR basis for obvious reasons (github comments)
+  if (isPR()) {
+    def benchmarkGitHubFile = 'report.md'
+    def benchmarkResults = 'benchmark-results'
+    def currentBenchmarkResults = "build/${benchmarkResults}"
+    def baseline = "build/${env.TARGET_BRANCH}/${benchmarkResults}"
```
Comment (v1v) [inline .ci/Jenkinsfile:164]:
```suggestion
    def baseline = "build/${env.CHANGE_TARGET}/${benchmarkResults}"
```

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -150,6 +155,56 @@ pipeline {
...
+        }
+      }
+
+      // for debugging purposes
+      if (fileExists(currentBenchmarkResults)) {
+        sh(label: 'debug current benchmark', script: "ls -l ${currentBenchmarkResults}", returnStatus: true)
+      }
+      if (fileExists(currentBenchmarkResults)) {
+        sh(label: 'debug baseline benchmark', script: "ls -l ${baseline}", returnStatus: true)
+      }
```
Comment (jsoriano) [inline .ci/Jenkinsfile:187]: Both of these ifs are checking the same (`fileExists(currentBenchmarkResults)`), is this expected?

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -150,6 +155,56 @@ pipeline {
...
+    def baseline = "build/${env.CHANGE_TARGET}/${benchmarkResults}"
+    dir("${BASE_DIR}") {
+      // download artifacts for this PR
+      def bucketUri = getBenchmarkBucketURI() + "*.xml"
+      googleStorageDownload(bucketUri: bucketUri, credentialsId: "${JOB_GCS_CREDENTIALS}", localDirectory: currentBenchmarkResults, pathPrefix: getBenchmarkPathPrefix())
+
+      // download artifacts from the target branch if any
+      if (env.CHANGE_TARGET) {
+        try {
+          projectName = env.JOB_NAME.replace(env.JOB_BASE_NAME, env.CHANGE_TARGET)
```
Comment (jsoriano) [inline .ci/Jenkinsfile:173]: I find a bit hard to follow how the benchmarks of the target branch are obtained, could you please add a comment here with a brief explanation? And maybe move this code to its own function.

File: dev/benchreport/report.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+type report struct {
+	Package    string
+	DataStream string
+	Old        float64
+	New        float64
+	Diff       float64
+	Percentage float64
+}
+
+func GetBenchReport(sourceDir, targetDir string, threshold float64) ([]byte, error) {
```
Comment (jsoriano) [inline dev/benchreport/report.go:54]: Nit. Mention the kind of report that is being generated here.

Something like:
```suggestion
func GetGithubMarkdownBenchReport(sourceDir, targetDir string, threshold float64) ([]byte, error) {
```

File: dev/benchreport/report.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	"fmt"
+	"math"
+	"os"
+	"text/template"
+)
+
+const (
+	reportExt = ".xml"
+	tpl       = `### :rocket: Benchmarks report
+{{range $package, $reports := .}}
```
Comment (jsoriano) [inline dev/benchreport/report.go:18]: Nit. This creates huge comments in Github with the number of integrations we have. Is there any way to have the report in jenkins, and link it from the Github comment when the number of reports is bigger than a certain number?

**Review by jsoriano — COMMENTED** [review]
Nice!

Just to confirm. In the comment with the report in https://github.com/elastic/integrations/pull/4205#issuecomment-1246532780, there seem to be a big variability of results, but it seems to be because they are using a dummy baseline?

File: .ci/Jenkinsfile (hunk 2 above)
Comment (v1v) [reply .ci/Jenkinsfile:187]: my bad :/

File: .ci/Jenkinsfile (hunk 3 above)
Comment (v1v) [reply .ci/Jenkinsfile:173]: This particular logic was coming from -> https://github.com/elastic/apm-pipeline-library/blob/54b4c9b43747d5a0c58bb8605f9f0c0d2d6510d5/vars/generateReport.groovy#L68-L89

But added to this pipeline so there is no need to use the quite opinionated approach with the shared library.

The idea is to create the full job name for a multibranch pipeline, by replacing the PR with the target branch to be able to download the baseline artifacts to compare with. I'll add a section

Comment (v1v) [reply .ci/Jenkinsfile]: See https://github.com/elastic/integrations/pull/4205/commits/fcd8d31af29f23ae09e8b13dd80f9ac8f316d626

File: dev/benchreport/report.go (hunk 5 above)
Comment (v1v) [reply dev/benchreport/report.go:18]: We discussed to provide a short report by default, and if needed to create the whole report by using some specific GitHub command, so it could be parsed in the CI and passed to the mage command, just an idea

Comment (marc-gr) [reply dev/benchreport/report.go:18]: working on this https://github.com/elastic/observability-robots/issues/1387#issuecomment-1246587189 atm, plus usually PRs are going to affect a single package, we should probably skip benchs (and tests?) if there is no diff for the package

File: .ci/Jenkinsfile [hunk 6]
```diff
@@ -51,7 +51,7 @@ pipeline {
     quietPeriod(10)
   }
   triggers {
-    issueCommentTrigger("${obltGitHubComments()}")
+    issueCommentTrigger("${obltGitHubComments()}|^/test benchmark fullreport")
```
Comment (v1v) [inline .ci/Jenkinsfile:54]: to support the existing default commands and the new one 
```suggestion
    issueCommentTrigger("(${obltGitHubComments()}|^/test benchmark fullreport)")
```

**Review by v1v — APPROVED** [review]
LGTM, really well done!

**Comment by jsoriano** [comment]
Discussed with Marc offline about reusing this in other projects. Agreed to merge this by now and continue evolving the solution with this in mind.

**Review by jsoriano — APPROVED** [review]

**Comment by ruflin** [comment]
Great to see this PR landed. Two asks from my end:

* How to use this is documented so everyone package dev can find it?
* All integrations dev are aware of this new feature and start using it
